### PR TITLE
Cell segmentation

### PR DESCRIPTION
**Pull request by MikeLippincott**
This PR uses nuclei and cell staining to segment cells from organoid microscopy images using the 2.5D approach.

### CODE REVIEW

File: 1.segment_images/scripts/1.segment_cells_organoids.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+original_cyto_image = cyto.copy()
+
+original_nuclei_z_count = nuclei.shape[0]
+original_cyto_z_count = cyto.shape[0]
+
+
+# In[5]:
+
+
+# make a 2.5 D max projection image stack with a sliding window of 3 slices
```
Comment (axiomcura) [inline 1.segment_images/scripts/1.segment_cells_organoids.py:116]: Make sure to add term that pertains to the cytoskeleton

File: 1.segment_images/scripts/1.segment_cells_organoids.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    image_stack_2_5D = np.append(
+        image_stack_2_5D, np.max(image_stack_window, axis=0)[np.newaxis, :, :], axis=0
+    )
+
+image_stack_2_5D = np.array(image_stack_2_5D)
+cyto = np.array(image_stack_2_5D)
+print("2.5D cyto image stack shape:", cyto.shape)
+
+
+# make a 2.5 D max projection image stack with a sliding window of 3 slices
```
Comment (axiomcura) [inline 1.segment_images/scripts/1.segment_cells_organoids.py:132]: Similar to the previous comment, this pertains specifically to the nuclei-related code. It appears to be duplicated—could this be refactored into a utility function for better reusability? If refactoring impacts downstream functionality, feel free to leave it as is.

File: 1.segment_images/scripts/1.segment_cells_organoids.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+diameter = 200
+
+masks_all_dict = {"masks": [], "imgs": []}
+imgs = np.array(imgs)
+
+for img in imgs:
+    masks, flows, styles, diams = model.eval(img, diameter=diameter, channels=channels)
+    masks_all_dict["masks"].append(masks)
+    masks_all_dict["imgs"].append(img)
+print(len(masks_all_dict))
```
Comment (axiomcura) [inline 1.segment_images/scripts/1.segment_cells_organoids.py:205]: Consider adding a comment here of what you are doing with the masks and images

File: 1.segment_images/scripts/3.make_cell_segmentation_videos.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env python
```
Comment (axiomcura) [inline 1.segment_images/scripts/3.make_cell_segmentation_videos.py:1]: Please consider re executing your notebook, the cells are not in order.

File: 1.segment_images/scripts/1.segment_cells_organoids.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+print(f"Decoupling the sliding window max projection of {window_size} slices")
+
+for z_stack_mask_index in range(len(masks_all)):
+    z_stack_decouple = []
+    [z_stack_decouple.append(masks_all[z_stack_mask_index]) for _ in range(window_size)]
+    for z_window_index, z_stack_mask in enumerate(z_stack_decouple):
+        if not (z_stack_mask_index + z_window_index) >= original_cyto_z_count:
+            reconstruction_dict[z_stack_mask_index + z_window_index].append(
+                z_stack_mask
+            )
```
Comment (axiomcura) [inline 1.segment_images/scripts/1.segment_cells_organoids.py:247]: I would recommend adding some code comments here.

File: 1.segment_images/scripts/3.make_cell_segmentation_videos.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+# In[5]:
+
+
+frames = [cell_mask[i] for i in range(cell_mask.shape[0])]
+
+# Write the frames to a GIF
+output_file_path = pathlib.Path(output_path / "cell_mask_output.gif")
+imageio.mimsave(
+    output_file_path, frames, duration=0.1, loop=10
+)  # duration is the time between frames in seconds
```
Comment (axiomcura) [inline 1.segment_images/scripts/3.make_cell_segmentation_videos.py:102]: Beautiful gifs btw c:

**Review by axiomcura — APPROVED** [review]
LGTM and also awesome gifs! Left some comments !

File: 1.segment_images/local_segmentation.sh [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+jupyter nbconvert --to=script --FilesWriter.build_directory=scripts/ notebooks/*.ipynb
+
+if [ "$NOTEBOOK" = True ]; then
+    cd notebooks/ || exit
+    papermill 0.segment_nuclei_organoids.ipynb 0.segment_nuclei_organoids.ipynb
+    papermill 1.segment_cells_organoids.ipynb 1.segment_cells_organoids.ipynb
+    papermill 2.make_nuclei_segmentation_videos.ipynb 2.make_nuclei_segmentation_videos.ipynb
+    papermill 3.make_cell_segmentation_videos.ipynb 3.make_cell_segmentation_videos.ipynb
+    cd ../ || exit
+fi
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 1.segment_images/local_segmentation.sh:17]: added

File: 1.segment_images/scripts/1.segment_cells_organoids.py [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+# Import dependencies
+import numpy as np
+import skimage
+import tifffile
+from cellpose import core, io, models
+
+io.logger_setup()
+from cellpose.io import imread
+from PIL import Image
+from skimage import io
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 1.segment_images/scripts/1.segment_cells_organoids.py:25]: implemented change

File: 1.segment_images/scripts/1.segment_cells_organoids.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+    if "405" in f:
+        nuclei = io.imread(f)
+    elif "488" in f:
+        cyto1 = io.imread(f)
+    elif "555" in f:
+        cyto2 = io.imread(f)
+    elif "640" in f:
+        cyto3 = io.imread(f)
+    else:
+        print(f"Unknown channel: {f}")
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 1.segment_images/scripts/1.segment_cells_organoids.py:100]: added

File: 1.segment_images/scripts/1.segment_cells_organoids.py (hunk 3 above)
Comment (MikeLippincott) [reply 1.segment_images/scripts/1.segment_cells_organoids.py:205]: added

File: 1.segment_images/scripts/3.make_cell_segmentation_videos.py (hunk 4 above)
Comment (MikeLippincott) [reply 1.segment_images/scripts/3.make_cell_segmentation_videos.py:1]: converted

File: 1.segment_images/scripts/1.segment_cells_organoids.py (hunk 5 above)
Comment (MikeLippincott) [reply 1.segment_images/scripts/1.segment_cells_organoids.py:247]: added
